File: packages/core/src/validation.ts
import type { AdapterType, ConnectionConfig } from './types/index.js';

export interface ValidationResult {
  valid: boolean;
  error?: string;
}

const VALID_ADAPTER_TYPES: AdapterType[] = [
  'postgres',
  'mysql',
  'sqlite',
  'mariadb',
  'mssql',
  'mongodb',
  'redis',
  'clickhouse',
];

export function validateConnectionConfig(
  type: AdapterType,
  config?: Partial<ConnectionConfig> | null
): ValidationResult {
  if (!type || !VALID_ADAPTER_TYPES.includes(type)) {
    return { valid: false, error: `Invalid or unsupported database type: '${type}'` };
  }

  if (!config || typeof config !== 'object') {
    return { valid: false, error: 'Connection configuration is required' };
  }

  if (type === 'sqlite') {
    const dbPath = (config.filepath || config.database || '').trim();
    if (!dbPath) {
      return { valid: false, error: 'Database file path is required for SQLite' };
    }
    return { valid: true };
  }

  if (type === 'mongodb' && config.filepath && config.filepath.trim().length > 0) {
    return { valid: true };
  }

  // Host validation
  if (config.host === undefined || config.host === null || typeof config.host !== 'string' || config.host.trim() === '') {
    return { valid: false, error: 'Host is required' };
  }

  // Port validation
  if (config.port === undefined || config.port === null || (typeof config.port === 'string' && (config.port as string).trim() === '')) {
    return { valid: false, error: 'Port is required' };
  }
  const portNum = typeof config.port === 'number' ? config.port : Number(config.port);
  if (isNaN(portNum) || !Number.isInteger(portNum) || portNum < 1 || portNum > 65535) {
    return { valid: false, error: 'Port must be a valid integer between 1 and 65535' };
  }

  // Database name validation
  if (type !== 'redis') {
    if (config.database === undefined || config.database === null || typeof config.database !== 'string' || config.database.trim() === '') {
      return { valid: false, error: 'Database name is required' };
    }
  }

  // Username validation
  if (type === 'postgres' || type === 'mysql' || type === 'mariadb' || type === 'mssql') {
    if (config.username === undefined || config.username === null || typeof config.username !== 'string' || config.username.trim() === '') {
      return { valid: false, error: 'Username is required' };
    }
  }

  return { valid: true };
}

export function assertValidConnectionConfig(
  type: AdapterType,
  config?: Partial<ConnectionConfig> | null
): void {
  const result = validateConnectionConfig(type, config);
  if (!result.valid) {
    throw new Error(result.error);
  }
}

File: crates/core/src/validation.rs
use crate::config::{AdapterType, ConnectionConfig};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ValidationResult {
    pub valid: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

pub fn validate_connection_config(
    adapter_type: AdapterType,
    config: Option<&ConnectionConfig>,
) -> ValidationResult {
    let config = match config {
        Some(c) => c,
        None => {
            return ValidationResult {
                valid: false,
                error: Some("Connection configuration is required".to_string()),
            };
        }
    };

    if adapter_type == AdapterType::Sqlite {
        let db_path = config
            .filepath
            .as_deref()
            .or(config.database.as_deref())
            .unwrap_or("")
            .trim();
        if db_path.is_empty() {
            return ValidationResult {
                valid: false,
                error: Some("Database file path is required for SQLite".to_string()),
            };
        }
        return ValidationResult {
            valid: true,
            error: None,
        };
    }

    if adapter_type == AdapterType::Mongodb {
        if let Some(fp) = &config.filepath {
            if !fp.trim().is_empty() {
                return ValidationResult {
                    valid: true,
                    error: None,
                };
            }
        }
    }

    // Host validation
    match &config.host {
        Some(h) if !h.trim().is_empty() => {}
        _ => {
            return ValidationResult {
                valid: false,
                error: Some("Host is required".to_string()),
            };
        }
    }

    // Port validation
    match config.port {
        Some(port) if port >= 1 => {}
        _ => {
            return ValidationResult {
                valid: false,
                error: Some("Port must be a valid integer between 1 and 65535".to_string()),
            };
        }
    }

    // Database name validation
    if adapter_type != AdapterType::Redis {
        match &config.database {
            Some(db) if !db.trim().is_empty() => {}
            _ => {
                return ValidationResult {
                    valid: false,
                    error: Some("Database name is required".to_string()),
                };
            }
        }
    }

    // Username validation
    if matches!(
        adapter_type,
        AdapterType::Postgres | AdapterType::Mysql | AdapterType::Mariadb | AdapterType::Mssql
    ) {
        match &config.username {
            Some(u) if !u.trim().is_empty() => {}
            _ => {
                return ValidationResult {
                    valid: false,
                    error: Some("Username is required".to_string()),
                };
            }
        }
    }

    ValidationResult {
        valid: true,
        error: None,
    }
}

pub fn assert_valid_connection_config(
    adapter_type: AdapterType,
    config: Option<&ConnectionConfig>,
) -> Result<(), String> {
    let result = validate_connection_config(adapter_type, config);
    if !result.valid {
        Err(result.error.unwrap_or_else(|| "Invalid connection config".to_string()))
    } else {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_missing_config() {
        let res = validate_connection_config(AdapterType::Postgres, None);
        assert!(!res.valid);
        assert_eq!(res.error.as_deref(), Some("Connection configuration is required"));
    }

    #[test]
    fn test_sqlite_valid_and_invalid() {
        let invalid = ConnectionConfig {
            filepath: Some("   ".to_string()),
            ..Default::default()
        };
        assert!(!validate_connection_config(AdapterType::Sqlite, Some(&invalid)).valid);

        let valid = ConnectionConfig {
            filepath: Some("/path/to/db.sqlite".to_string()),
            ..Default::default()
        };
        assert!(validate_connection_config(AdapterType::Sqlite, Some(&valid)).valid);
    }

    #[test]
    fn test_postgres_validation() {
        let mut config = ConnectionConfig {
            host: Some("localhost".to_string()),
            port: Some(5432),
            database: Some("mydb".to_string()),
            username: Some("postgres".to_string()),
            ..Default::default()
        };
        assert!(validate_connection_config(AdapterType::Postgres, Some(&config)).valid);

        config.username = None;
        assert!(!validate_connection_config(AdapterType::Postgres, Some(&config)).valid);
    }

    #[test]
    fn test_redis_validation() {
        let config = ConnectionConfig {
            host: Some("localhost".to_string()),
            port: Some(6379),
            database: None, // Redis doesn't require database name
            ..Default::default()
        };
        assert!(validate_connection_config(AdapterType::Redis, Some(&config)).valid);
    }
}
